src/dispatch.rs

```diff
@@ -0,0 +1,103 @@
+use crate::idbstore::IdbStore;
+use async_std::sync::{channel, Receiver, Sender};
+use std::collections::HashMap;
+use std::sync::Mutex;
+use wasm_bindgen_futures::spawn_local;
+
+type Response = Result<String, String>;
+
+struct Request {
+    db_name: String,
+    rpc: String,
+    data: String,
+    response: Sender<Response>,
+}
+
+lazy_static! {
+    static ref SENDER: Mutex<Sender::<Request>> = {
+        let (tx, rx) = channel::<Request>(1);
+        spawn_local(dispatch_loop(rx));
+        Mutex::new(tx)
+    };
+}
+
+async fn dispatch_loop(rx: Receiver<Request>) {
+    let mut dispatcher = Dispatcher {
+        connections: HashMap::new(),
+    };
+
+    loop {
+        match rx.recv().await {
+            Err(why) => log!("Recv failed: {}", why),
+            Ok(req) => {
+                let response = match req.rpc.as_str() {
+                    "open" => dispatcher.open(&req).await,
+                    "close" => dispatcher.close(&req).await,
+                    "debug" => dispatcher.debug(&req).await,
+                    _ => Err("Unsupported rpc name".to_string()),
+                };
+                req.response.send(response).await;
+            }
+        }
+    }
+}
+
+struct Dispatcher {
+    connections: HashMap<String, IdbStore>,
+}
+
+impl Dispatcher {
+    async fn open(&mut self, req: &Request) -> Response {
+        if req.db_name.len() == 0 {
+            return Err("db_name must be non-empty".to_string());
+        }
+        if self.connections.contains_key(&req.db_name[..]) {
+            return Ok("".to_string());
+        }
+        match IdbStore::new(&req.db_name[..]).await {
+            Err(v) => {
+                log!("Failed to open! {:?}", v);
+                return Err("Failed to open failed!".to_string());
+            }
+            Ok(v) => {
+                if let Some(v) = v {
+                    self.connections.insert(req.db_name.clone(), v);
+                }
+            }
+        }
+        Ok("".to_string())
+    }
+
+    async fn close(&mut self, req: &Request) -> Response {
+        if !self.connections.contains_key(&req.db_name[..]) {
+            return Ok("".to_string());
+        }
+        self.connections.remove(&req.db_name);
+
+        Ok("".to_string())
+    }
+
+    async fn debug(&mut self, req: &Request) -> Response {
+        match req.data.as_str() {
+            "open_dbs" => Ok(format!("{:?}", self.connections.keys())),
+            _ => Err("Debug command not defined".to_string()),
+        }
+    }
+}
+
+pub async fn dispatch(db_name: String, rpc: String, data: String) -> Response {
+    let (tx, rx) = channel::<Response>(1);
+    let request = Request {
+        db_name,
+        rpc,
+        data,
+        response: tx,
+    };
+    {
+        SENDER.lock().unwrap().send(request).await;
```

Comment:
Avoid `unwrap`. Instead log the error?

Reply:
In this case the Result is only to protect against the mutex having been "poisoned" (see https://doc.rust-lang.org/std/sync/struct.Mutex.html), meaning that a holder of it panicked. You can test for this and access the value without acquiring the mutex, but I think at that point the error is unrecoverable, and there's no way to put the mutex back in a good state.

In such cases, do you think we should go through the trouble to trying to explicitly report this (the triggering panic should have already been reported) or just leave this be? My vote is to just `unwrap()` here, but I agree that it's too bad that these semantics make code reviewing unwrap() calls tricky since it's sometimes verbotten and sometimes boilerplate.

Reply:
We talked about this early on and came to this consensus:

> * If possible, do not intentionally crash processes (eg, by panicing on unexpected conditions).
>    * For our client-side software, this is bad practice because we frequently run in-process with our users and we will crash them too!

-- https://github.com/rocicorp/replicache/blob/master/contributing.md#style-general

I think the core consideration is that we do not ever want to crash our host software.

However. I do not know exactly what panic and crash mean in wasm. I assume (?) that if a wasm program panics the wasm module goes away but the webpage is otherwise unaffected. In that case, perhaps we don't care.

But then we should think about what it means to panic in other environments.

If we can arrange for a global panic handler in Rust to convert panics into logs that don't crash the host program in the environments we care about, then I agree that would be much more convenient than meticulously avoiding panics.

cc @phritz

Reply:
I changed this call, which ended up being easy. I just wanted to say that I agree with the sentiment above, but this is still a weird case since Mutex.lock() is defined as not being allowed to error unless another holder of it has already panicked and left the mutex in a permanently poisoned state.

Btw, I did a little research on trying to avoid panics in Rust, and it sounds like it's considered to be a hard problem because they're so baked in to the standard library. I did find a crate (https://crates.io/crates/no-panic/0.1.5) with a macro that forces the compiler to prove that a function wont panic, but it's incompatible with async functions. :( Googling for "Rust no_panic" yields a bunch of interesting reading, however.